pubnub/event_engine/effects.py

```diff
@@ -0,0 +1,431 @@
+            self.event_engine.trigger(events.HeartbeatGiveUpEvent(channels=self.invocation.channels,
+                                                                  groups=self.invocation.groups,
+                                                                  reason=self.invocation.reason,
+                                                                  attempt=self.invocation.attempts))
+        request = Heartbeat(self.pubnub).channels(channels).channel_groups(groups).cancellation_event(stop_event)
+        delay = self.calculate_reconnection_delay(attempt)
+        self.logger.warning(f'Will retry to Heartbeat in {delay}s')
+        await asyncio.sleep(delay)
+
+        response = await request.future()
+        if isinstance(response, PubNubException):
+            self.logger.warning(f'Heartbeat failed: {str(response)}')
+            self.event_engine.trigger(events.HeartbeatFailureEvent(channels=channels, groups=groups,
+                                                                   reason=response.status.error_data,
+                                                                   attempt=attempt))
+        elif response.status.error:
+            self.logger.warning(f'Heartbeat failed: {response.status.error_data.__dict__}')
+            self.event_engine.trigger(events.HeartbeatFailureEvent(channels=channels, groups=groups,
+                                                                   reason=response.status.error_data,
+                                                                   attempt=attempt))
+        else:
+            self.event_engine.trigger(events.HeartbeatSuccessEvent(channels=channels, groups=groups))
+
+
+class EffectFactory:
+    _managed_invocations = {
+        invocations.HandshakeInvocation.__name__: HandshakeEffect,
+        invocations.ReceiveMessagesInvocation.__name__: ReceiveMessagesEffect,
+        invocations.HandshakeReconnectInvocation.__name__: HandshakeReconnectEffect,
+        invocations.ReceiveReconnectInvocation.__name__: ReceiveReconnectEffect,
+        invocations.HeartbeatInvocation.__name__: HeartbeatEffect,
+        invocations.HeartbeatWaitInvocation.__name__: HeartbeatWaitEffect,
+        invocations.HeartbeatDelayedHeartbeatInvocation.__name__: HeartbeatDelayedEffect,
+        invocations.HeartbeatLeaveInvocation.__name__: HeartbeatLeaveEffect,
+    }
+
+    def __init__(self, pubnub_instance, event_engine_instance) -> None:
+        self._pubnub = pubnub_instance
+        self._event_engine = event_engine_instance
+
+    def create(self, invocation: invocations.PNInvocation) -> Effect:
+        if invocation.__class__.__name__ not in self._managed_invocations:
+            raise PubNubException(errormsg=f"Unhandled Invocation: {invocation.__class__.__name__}")
+        return self._managed_invocations[invocation.__class__.__name__](self._pubnub, self._event_engine, invocation)
+
+
+class EmitEffect:
```

Comment:
There should be two separate Effects for emitting messages and status, respectively. Let's leave it if this requires significant amount of work